src/main/scala/mesosphere/marathon/core/storage/impl/CachingPersistenceStore.scala

```diff
@@ -0,0 +1,158 @@
+package mesosphere.marathon.core.storage.impl
+
+import akka.http.scaladsl.marshalling.Marshaller
+import akka.http.scaladsl.unmarshalling.Unmarshaller
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import akka.{Done, NotUsed}
+import com.typesafe.scalalogging.StrictLogging
+import mesosphere.marathon.{PrePostDriverCallback, StoreCommandFailedException}
+import mesosphere.marathon.core.storage.{IdResolver, PersistenceStore}
+import mesosphere.marathon.util.toRichFuture
+
+import scala.async.Async.{async, await}
+import scala.collection.concurrent.TrieMap
+import scala.concurrent.{Future, Promise}
+import scala.util.{Failure, Success, Try}
+
+/**
+  * A Write Ahead Cache of another persistence store that preloads the entire persistence store into memory before
+  * satisfying any requests.
+  *
+  * TODO: Consider an alternative strategy where we see if the promise is complete and use it
+  * otherwise going directly to the storage layer. This turns out to be much more complicated
+  * as the cache is populated asynchronously, so it would have to queue up all create/update operations
+  * onto the future to keep the value fully updated: then, there would be a short window of time when
+  * the cached data is actually stale.
+  *
+  * @param store The store to cache
+  * @param keyAsPath Conversion from the Persistence layer's key to a stringified path.
+  * @param mat a materializer for akka streaming
+  * @tparam Serialized The serialized format for the persistence store.
+  */
+class CachingPersistenceStore[K, Serialized](
+    store: PersistenceStore[K, Serialized],
+    keyAsPath: K => String,
+    pathAsKey: String => K)(
+    implicit
+    override val mat: Materializer
+) extends PersistenceStore[K, Serialized] with StrictLogging with PrePostDriverCallback {
+
+  private[storage] val cache = Promise[TrieMap[String, Either[Serialized, Any]]]()
+
+  override def preDriverStarts: Future[Unit] = {
+    val storage = TrieMap[String, Either[Serialized, Any]]()
+    val future = store.keys().mapAsync(8) { key =>
+      store.rawGet(key).map(v => key -> v)
+    }.runForeach {
+      case (key, maybeSerialized) =>
+        maybeSerialized match {
+          case Some(v) =>
+            storage.put(keyAsPath(key), Left(v))
+          case None =>
+        }
+    }
+    cache.completeWith(future.map(_ => storage))
```

Comment:
`preDriverStart` will be called every time we got elected (multiple times). The promise here can be completed only once. Perhaps create a new Promise for every call on `preDriverStarts`?